packages/mobile/src/harmony-native/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,311 @@
+        onBlurProp?.(e)
+      },
+      [focused, motion.press, onBlurProp]
+    )
+
+    const [isFocusedState, onFocus, onBlur] = useFocusState(
+      handleFocus,
+      handleBlur
+    )
+
+    const isFocused = _isFocused ?? isFocusedState
+
+    // For focus behavior and accessiblity, <label> needs to have a htmlFor={} provided to an id matching the input
+    // const backupId = useId()
+    const id = idProp // ?? backupId
+
+    const characterCount = value !== undefined ? `${value}`.length : 0
+    const hasValue = characterCount > 0
+
+    // Hide the label when requested or when the size is set to small
+    const shouldShowLabel = !hideLabel && size !== TextInputSize.SMALL
+    const labelText = required ? `${labelProp} *` : labelProp
+    const placeholderText =
+      required && hideLabel ? `${placeholder} *` : placeholder
+    const helperTextSize: TextSize = size === TextInputSize.SMALL ? 'xs' : 's'
+
+    // Whenever a label isn't visible the placeholder should be visible in it's place (if provided)
+    const shouldShowPlaceholder = isFocused || !shouldShowLabel
+    const shouldShowAdornments = isFocused || hasValue || !shouldShowLabel
+    // Show the maxlength text whenever we hit a certain threshold (default 70%) + the input is focused
+    const shouldShowMaxLengthText =
+      isFocused &&
+      maxLength &&
+      characterCount >= showMaxLengthThreshold * maxLength
+    // Turn the maxlength text to the warning color whenever we hit a certain threshold (default 90%)
+    let maxLengthTextColor: TextColors = 'default'
+    if (maxLength && characterCount > maxLength) {
+      maxLengthTextColor = 'danger'
+    } else if (
+      maxLength &&
+      characterCount >= maxLengthWarningThreshold * maxLength
+    ) {
+      maxLengthTextColor = 'warning'
+    }
+
+    const isSmall = size === TextInputSize.SMALL
+    const stateColor = disabled
+      ? 'subdued'
+      : error
+      ? 'danger'
+      : warning || _incorrectError
+      ? 'warning'
+      : undefined
+
+    const handlePressRoot = useCallback(() => {
+      if (!isFocused) {
+        innerInputRef.current?.focus()
+      }
+    }, [isFocused])
+
+    const tap = Gesture.Tap()
+      .onBegin(() => {
+        focused.value = withTiming(1, motion.press)
+      })
+      .onFinalize((event) => {
+        if (event.state !== State.END) {
+          focused.value = withTiming(0, motion.press)
+        }
+      })
+
+    const animatedRootStyles = useAnimatedStyle(() => {
+      const staticBorderColor = _isFocused
+        ? color.secondary.s400
+        : stateColor
+        ? color.text[stateColor]
+        : undefined
+
+      return {
+        borderColor:
+          staticBorderColor ??
+          interpolateColor(
+            focused.value,
+            [0, 1],
+            [color.border.default, color.secondary.s400]
+          )
+      }
+    })
+
+    const animatedLabelStyle = useAnimatedStyle(() => ({
```

Comment:
this animation logic looks much simpler than the existing stuff I messed with earlier today haha 💯

Reply:
it def got a bit wild, but this part at least is clean